tests/splinepyTest.py

```diff
@@ -0,0 +1,581 @@
+        self.z = splinepy.Bezier(**self.z2P2D)
+
+        # set up queries
+        self.q2D = [
+            [.01, .01], 
+            [.01, .5], 
+            [.9, .1], 
+            [.8, .7], 
+            [.4, .99],
+        ]
+
+    def test_constructor_degree_check(self):
+        """ Test the raise Error for mismatch of degree and par_dim. """
+        # not for Bezier
+
+        self.b2P2D.update(degrees = [2,])
+        self.n2P2D.update(degrees = [2,])
+
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.BSpline(**self.b2P2D)
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.NURBS(**self.n2P2D)
+
+    def test_constructor_kv_nr_check(self):
+        """ Test the raise Error for missing knot vectors. """
+        # not for Bezier
+
+        self.b2P2D.update(knot_vectors = [[0, 0, 0, 0.5, 1, 1, 1]])
+        self.n2P2D.update(knot_vectors = [[0, 0, 0, 1, 1, 1]])
+
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.BSpline(**self.b2P2D)
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.NURBS(**self.n2P2D)
+
+    def test_constructor_kv_length_check(self):
+        """ Test the raise Error for length of knot vectors. """
+        # not for Bezier
+
+        self.b2P2D.update(knot_vectors = [
+            [0, 0, 0.5, 1, 1], 
+            [0, 0, 1, 1]
+            ]
+        )
+        self.n2P2D.update(knot_vectors = [
+            [0, 0, 1, 1], 
+            [0, 0, 1, 1]
+            ]
+        )
+
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.BSpline(**self.b2P2D)
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.NURBS(**self.n2P2D)
+
+    def test_constructor_cp_nr_check(self):
+        """ Test the raise Error for invalid number of control points. """
+        # not for Bezier
+
+        self.b2P2D.update(
+            control_points = [
+                [0, 0],
+                [0, 1],
+                [1, 1.5],
+                [3, 1.5],
+                [-2, 0],
+                [-2, 2],
+                [1, 5],
+                [3, 5],
+                ]
+        )
+        self.n2P2D.update(
+            control_points = [
+                [-1.,  0.],
+                [ 0.,  1.],
+                [-2.,  0.],
+                [0.,  2.],
+                ],
+            weights = [
+                [1.],
+                [1.],
+                [1.],
+                [1.],
+                ],         
+        )
+        
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.BSpline(**self.b2P2D)
+        with self.assertRaises(splinepy._spline.InputDimensionError) as context:
+            splinepy.NURBS(**self.n2P2D)
+
+    def test_basis_functions(self):
```

Comment:
Left the hard comparison in for now, analytical solution to be added, see #33. Added check of partition of unity of the basis functions from BSpline and NURBS as this is currently not possible for Beziers #32

Reply:
I updated #33. I think this comment was a bit to fuzzy... What I meant was by no means the implementation of analytical solutions (e.g. in the style of sympy). ImO, hard coded comparisons should always be avoided as they only ever test the functions at precice values. 
 
I rather meant something like this:
```
simple_bspline = sp.BSpline(
    degrees=[2],
    control_points=[
        [0, 0],
        [1, 1],
        [2, 0]
    ],
    knot_vectors=[[0,0,0,1,1,1]]
)
# Define quadratic basis functions
def basis_function(x):
    return np.array([
        (1-x)**2,
        2*x*(1-x),
        x**2
    ])
# Make a comparison at arbitrary values
x = np.random.rand()

self.assertAlmostEqual(simple_bspline.basis_function(x), basis_function(x))
```

**THIS IS JUST AN EXAMPLE AND NOT A PROPOSAL**
But it could serve as inspiration when designing the tests. ImO, generality is key here. You must find creative solutions to be able to test at arbitrary values with as little hard-coding as possible, e.g., you could test the partition of unity at random points, or if the non-zero basis functions are non-zero (and only them), etc.